### ISSUE TITLE
Import errors when running with Deno

### ISSUE THREAD

**Issue by dsego** [nanojsx/nano#3]
I am importing Nano as shown on the website to use as an SSR library.
```js
import * as Nano from 'https://deno.land/x/nano_jsx@v0.0.11/mod.ts'
```
This is my `tsconfig.json`
```json
{
  "compilerOptions": {
    "target": "ESNext",
    "jsx": "react",
    "jsxFactory": "Nano.h"
  }
}     

```

But it seems that Nano wants browser stuff and Deno just gives up with 67 errors :pensive:

<details>
<summary>
Click here to see the output.
</summary>

```
error: TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
  const tmpNode = document.createElement('div')
                  ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/helpers.ts:2:19

TS2304 [ERROR]: Cannot find name 'MutationObserver'.
  var observer = new MutationObserver((mutationsList) => {
                     ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/helpers.ts:13:22

TS7006 [ERROR]: Parameter 'mutationsList' implicitly has an 'any' type.
  var observer = new MutationObserver((mutationsList) => {
                                       ~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/helpers.ts:13:40

TS7006 [ERROR]: Parameter 'mutation' implicitly has an 'any' type.
    mutationsList.forEach((mutation) => {
                           ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/helpers.ts:14:28

TS7006 [ERROR]: Parameter 'removed' implicitly has an 'any' type.
      mutation.removedNodes.forEach((removed) => {
                                     ~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/helpers.ts:15:38

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
  observer.observe(document, {
                   ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/helpers.ts:26:20

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
export const removeAllChildNodes = (parent: HTMLElement) => {
                                            ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:17:45

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
      let c = renderComponent(child) as HTMLElement
                                        ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:38:41

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
      else element.appendChild(c.nodeType == null ? document.createTextNode(c.toString()) : c)
                                                    ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:42:53

TS2304 [ERROR]: Cannot find name 'SVGElement'.
  const child = props.children[0] as SVGElement
                                     ~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:51:38

TS2304 [ERROR]: Cannot find name 'SVGElement'.
  const svg = hNS('svg') as SVGElement
                            ~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:54:29

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
export const hydrate = (component: any, parent: HTMLElement | null = null, removeChildNodes = true) => {
                                                ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:63:49

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
export const render = (component: any, parent: HTMLElement | null = null, removeChildNodes = true) => {
                                               ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:68:48

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
  if (el.component) return renderComponent(el) as HTMLElement
                                                  ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:156:51

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
  return document.createElementNS('http://www.w3.org/2000/svg', tag) as SVGElement
         ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:161:10

TS2304 [ERROR]: Cannot find name 'SVGElement'.
  return document.createElementNS('http://www.w3.org/2000/svg', tag) as SVGElement
                                                                        ~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:161:73

TS2304 [ERROR]: Cannot find name 'SVGElement'.
      ? (hNS('svg') as SVGElement)
                       ~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:174:24

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
      : (document.createElement(tagNameOrComponent) as HTMLElement)
         ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:175:10

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
      : (document.createElement(tagNameOrComponent) as HTMLElement)
                                                       ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/core.ts:175:56

TS2304 [ERROR]: Cannot find name 'HTMLCollection'.
  private _elements: HTMLCollection
                     ~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/component.ts:11:22

TS2304 [ERROR]: Cannot find name 'HTMLCollection'.
  public get elements(): HTMLCollection {
                         ~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/component.ts:48:26

TS2304 [ERROR]: Cannot find name 'HTMLCollection'.
  public set elements(e: HTMLCollection) {
                         ~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/component.ts:52:26

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
    const parent = nodeElements[0].parentElement as HTMLElement
                                                    ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/component.ts:108:53

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
    rendered.forEach((r: HTMLElement) => {
                         ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/component.ts:111:26

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
    nodeElements.forEach((t: HTMLElement) => {
                             ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/component.ts:118:30

TS2304 [ERROR]: Cannot find name 'IntersectionObserver'.
    const observer = new IntersectionObserver(
                         ~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/visible.ts:8:26

TS7006 [ERROR]: Parameter 'entries' implicitly has an 'any' type.
      (entries, observer) => {
       ~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/visible.ts:9:8

TS7006 [ERROR]: Parameter 'observer' implicitly has an 'any' type.
      (entries, observer) => {
                ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/visible.ts:9:17

TS7006 [ERROR]: Parameter 'entry' implicitly has an 'any' type.
        entries.forEach((entry) => {
                         ~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/visible.ts:10:26

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
    this.props.children.forEach((element: HTMLElement) => {
                                          ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:30:43

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
      let parent = this.props.footer ? document.body : document.head
                                       ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:31:40

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
      let parent = this.props.footer ? document.body : document.head
                                                       ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:31:56

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
        const htmlTag = document.getElementsByTagName(tag)[0]
                        ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:46:25

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
        const titleTags = document.getElementsByTagName('TITLE') as HTMLCollectionOf<HTMLTitleElement>
                          ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:52:27

TS2304 [ERROR]: Cannot find name 'HTMLCollectionOf'.
        const titleTags = document.getElementsByTagName('TITLE') as HTMLCollectionOf<HTMLTitleElement>
                                                                    ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:52:69

TS2304 [ERROR]: Cannot find name 'HTMLTitleElement'.
        const titleTags = document.getElementsByTagName('TITLE') as HTMLCollectionOf<HTMLTitleElement>
                                                                                     ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:52:86

TS2304 [ERROR]: Cannot find name 'HTMLTitleElement'.
          let e = element as HTMLTitleElement
                             ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:54:30

TS2304 [ERROR]: Cannot find name 'HTMLTitleElement'.
          let titleTag = h('title', null, element.innerHTML) as HTMLTitleElement
                                                                ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:57:65

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
      const el: HTMLElement[] = document.getElementsByTagName(tag) as any
                ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:67:17

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
      const el: HTMLElement[] = document.getElementsByTagName(tag) as any
                                ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:67:33

TS2339 [ERROR]: Property 'document' does not exist on type 'Window & typeof globalThis'.
    const isSSR = !(typeof window !== 'undefined' && window.document)
                                                            ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/helmet.ts:90:61

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
export const hydrateLazy = (component: any, parent: HTMLElement | null = null, removeChildNodes = true) => {
                                                    ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/lazy.ts:4:53

TS2304 [ERROR]: Cannot find name 'localStorage'.
    if (storage === 'memory' || typeof localStorage === 'undefined') return
                                       ~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/store.ts:22:40

TS2304 [ERROR]: Cannot find name 'localStorage'.
    const Storage = storage === 'local' ? localStorage : sessionStorage
                                          ~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/store.ts:24:43

TS2304 [ERROR]: Cannot find name 'sessionStorage'.
    const Storage = storage === 'local' ? localStorage : sessionStorage
                                                         ~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/store.ts:24:58

TS2304 [ERROR]: Cannot find name 'localStorage'.
    const Storage = this._storage === 'local' ? localStorage : sessionStorage
                                                ~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/store.ts:37:49

TS2304 [ERROR]: Cannot find name 'sessionStorage'.
    const Storage = this._storage === 'local' ? localStorage : sessionStorage
                                                               ~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/store.ts:37:64

TS2304 [ERROR]: Cannot find name 'HTMLImageElement'.
  image: HTMLImageElement
         ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/img.ts:6:10

TS2304 [ERROR]: Cannot find name 'IntersectionObserver'.
    const observer = new IntersectionObserver(
                         ~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/img.ts:11:26

TS7006 [ERROR]: Parameter 'entries' implicitly has an 'any' type.
      (entries, observer) => {
       ~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/img.ts:12:8

TS7006 [ERROR]: Parameter 'observer' implicitly has an 'any' type.
      (entries, observer) => {
                ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/img.ts:12:17

TS7006 [ERROR]: Parameter 'entry' implicitly has an 'any' type.
        entries.forEach((entry) => {
                         ~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/img.ts:13:26

TS2304 [ERROR]: Cannot find name 'HTMLImageElement'.
            this.image = h('img', { ...rest }) as HTMLImageElement
                                                  ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/img.ts:16:51

TS2304 [ERROR]: Cannot find name 'HTMLImageElement'.
    if (typeof lazy === 'boolean' && lazy === false) return h('img', { src, ...rest }) as HTMLImageElement
                                                                                          ~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/img.ts:32:91

TS2304 [ERROR]: Cannot find name 'IntersectionObserver'.
    const observer = new IntersectionObserver(
                         ~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:17:26

TS7006 [ERROR]: Parameter 'entries' implicitly has an 'any' type.
      (entries, observer) => {
       ~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:18:8

TS7006 [ERROR]: Parameter 'observer' implicitly has an 'any' type.
      (entries, observer) => {
                ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:18:17

TS7006 [ERROR]: Parameter 'entry' implicitly has an 'any' type.
        entries.forEach((entry) => {
                         ~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:19:26

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
    const links = document.getElementsByTagName('link')
                  ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:35:19

TS2304 [ERROR]: Cannot find name 'HTMLElement'.
      const prefetch = h('link', { rel: 'prefetch', href: this.props.href, as: 'document' }) as HTMLElement
                                                                                                ~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:44:97

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
      document.head.appendChild(prefetch)
      ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:45:7

TS2304 [ERROR]: Cannot find name 'HTMLLinkElement'.
        const target = e.target as HTMLLinkElement
                                   ~~~~~~~~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:55:36

TS2584 [ERROR]: Cannot find name 'document'. Do you need to change your target library? Try changing the `lib` compiler option to include 'dom'.
        if (target.href === document.referrer) window.history.back()
                            ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:56:29

TS2339 [ERROR]: Property 'history' does not exist on type 'Window & typeof globalThis'.
        if (target.href === document.referrer) window.history.back()
                                                      ~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:56:55

TS2339 [ERROR]: Property 'location' does not exist on type 'Window & typeof globalThis'.
        else window.location.href = target.href
                    ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:57:21

TS2339 [ERROR]: Property 'location' does not exist on type 'Window & typeof globalThis'.
        setTimeout(() => (window.location.href = href), delay)
                                 ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:63:34

TS2339 [ERROR]: Property 'document' does not exist on type 'Window & typeof globalThis'.
    if (prefetch === true && !(typeof window !== 'undefined' && window.document)) {
                                                                       ~~~~~~~~
    at https://deno.land/x/nano_jsx@v0.0.11/deno_lib/components/link.ts:84:72

Found 67 errors.
```
</details>

Deno version:
```
deno 1.5.2 (71d7482, release, x86_64-unknown-linux-gnu)
v8 8.7.220.3
typescript 4.0.5
```

**Comment by yandeu** [comment]
I have just published a private test repo. Hope it helps: https://github.com/nanojsx/nano-jsx-deno-dev

If it does not work, I will have a look this evening.

**Comment by yandeu** [comment]
Btw, I plan to release version 12 soon with a lot of new features and improvements. I will also try to write better docs.

**Comment by dsego** [comment]
Ok, thank you so much. Seems like your test repo is working fine. I've changed my tsconfig to match the test repo like so:
```json
{
  "compilerOptions": {
    "lib": ["dom", "dom.iterable", "esnext", "deno.ns", "deno.unstable"],
    "strictPropertyInitialization": false,
    "jsxFactory": "Nano.h"
  }
}     
```

And this seems to solve the Nano issues, but causes issues when importing Oak `import * as oak from "https://deno.land/x/oak/mod.ts";`

```
error: TS2339 [ERROR]: Property 'getIterator' does not exist on type 'ReadableStream<R>'.
  return res.readable.getIterator();
                      ~~~~~~~~~~~
    at https://deno.land/std@0.80.0/async/pool.ts:45:23
```

In any case, I can wait for the new release, thanks! :+1:

**Comment by yandeu** [comment]
I'm not sure what I should change to make it work. 

I have added oak to the [nano-jsx-deno-dev](https://github.com/nanojsx/nano-jsx-deno-dev) repo with the `--no-check` flag.

**Comment by yandeu** [comment]
Adding this seems to be even better:

```ts
// fixes an issue in std@0.80.0
interface ReadableStream<R> {
  getIterator(): any
}
```

**Comment by dsego** [comment]
@yandeu Thanks! :100:
The tsconfig and typings solve the importing issues for me. I see no reason not to close this ticket now :ok_hand:.

**Comment by yandeu** [comment]
Check out the latest deno example at https://github.com/nanojsx/nano-jsx-deno-example

**Comment by 1antares1** [comment]
> Adding this seems to be even better:
> 
> ```ts
> // fixes an issue in std@0.80.0
> interface ReadableStream<R> {
>   getIterator(): any
> }
> ```

Yeah! It worked exactly like this for me, just putting the other type to avoid ESLint alerts.
```typescript
 interfaz ReadableStream <R> {
     getIterator (): AsyncIterableIterator <R>;
 }
 ```
